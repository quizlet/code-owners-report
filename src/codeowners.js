// @flow
// Lots copied from https://github.com/beaugunderson/codeowners/blob/master/codeowners.js

import findUp from 'find-up'
import fs from 'fs'
import ignore from 'ignore'
import path from 'path'
import trueCasePath from 'true-case-path'

export type OwnersEntry = {
  path: string, // .gitignore style
  usernames: string[],
  match: string => boolean,
}

export const findCodeownersPath = (): string => {
  const codeownersPath = findUp.sync('CODEOWNERS', { cwd: process.cwd() })

  if (!codeownersPath) {
    throw new Error('findCodeownersPath() failed. Please specify instead.')
  }

  const trueCaseCodeownersPath = trueCasePath(codeownersPath)

  const codeownersFile = path.basename(trueCaseCodeownersPath)

  if (codeownersFile !== 'CODEOWNERS') {
    throw new Error(
      `Found a CODEOWNERS file but it was lower-cased: ${trueCaseCodeownersPath}`,
    )
  }
  return trueCaseCodeownersPath
}

function ownerMatcher(pathString) {
  const matcher = ignore().add(pathString)

  return function(fileString) {
    return matcher.ignores(fileString)
  }
}

export const parseCodeownersFile = (pathname: string): OwnersEntry[] => {
  const lines = fs
    .readFileSync(pathname)
    .toString()
    .split('\n')
  const ownerEntries = []

  lines.forEach(line => {
    if (!line) {
      return
    }

    if (line.startsWith('#')) {
      return
    }

    const [pathString, ...usernames] = line.split(/\s+/)

    ownerEntries.push({
      path: pathString,
      usernames,
      match: ownerMatcher(pathString),
    })
  })
  return ownerEntries
}

/*
 * Return list of owners per CODEOWNERS for the given file.
 */
// TODO memoize to cache between repeated calls from different sites
export const whoOwns = (
  ownersEntries: OwnersEntry[],
  filePath: string,
): string[] => {
  let lastMatchingEntry = { usernames: [] }
  for (const entry of ownersEntries) {
    if (entry.match(filePath)) {
      lastMatchingEntry = entry
    }
  }
  return lastMatchingEntry.usernames
}

export const addOwners = (
  ownersEntries: OwnersEntry[],
  filePath: string,
  metrics: Object,
) => {
  const owners = whoOwns(ownersEntries, filePath).join(', ')
  Object.assign(metrics, { owners })
}
