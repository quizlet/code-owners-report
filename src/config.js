// @flow

export type FileMetrics = {
  owners?: string[], // HACK to pass along owner info
  [measure: string]: number | boolean,
}

export type FilesMetricsMap = { [filename: string]: FileMetrics }

export type ExpressionMap = {
  [key: string]: RegExp,
}

export type RuleFlags = { [ruleId: string]: boolean }

export type ReportSpec = {
  omit?: RegExp,
  regexpMetrics?: ExpressionMap,
  eslintFlags?: RuleFlags,
}

export const FILES_SUM_METRIC_KEY = 'files count'
