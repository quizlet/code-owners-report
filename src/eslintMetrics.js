// @flow

import { CLIEngine } from 'eslint'

import type { RuleFlags } from './config'

const eslinter = new CLIEngine({
  useEslintrc: true,
})

type Counter = boolean => number

export const countEslintRuleViolations = (
  contents: string,
  ruleFlags: RuleFlags,
): $ObjMap<RuleFlags, Counter> => {
  const { results } = eslinter.executeOnText(contents)

  // only one result possible for executeOnText()
  const fileMessages = results[0].messages

  const eslintCounts = fileMessages.reduce((acc = {}, message) => {
    const { ruleId } = message
    if (ruleFlags[ruleId]) {
      acc[ruleId] = 1 + (acc[ruleId] || 0)
    }
    return acc
  }, {})

  return eslintCounts
}
