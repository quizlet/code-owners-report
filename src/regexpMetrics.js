// @flow

import type { ExpressionMap } from './config'

type Tester = RegExp => boolean

export const testRegExpMatches = (
  contents: string,
  expressionMap: ExpressionMap,
): $ObjMap<ExpressionMap, Tester> => {
  const matches = {}
  for (const key of Object.keys(expressionMap)) {
    const re = expressionMap[key]
    matches[key] = !!contents.match(re)
  }
  return matches
}
