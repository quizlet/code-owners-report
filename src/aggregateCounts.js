// @flow

import type { FilesMetricsMap, ReportSpec } from './config'

import { whoOwns, type OwnersEntry } from './codeowners'

import { FILES_SUM_METRIC_KEY } from './config'

const mkInitialCounts = (spec: ReportSpec) => {
  const headings = [FILES_SUM_METRIC_KEY].concat(
    spec.regexpMetrics && Object.keys(spec.regexpMetrics),
    spec.eslintFlags && Object.keys(spec.eslintFlags),
  )
  const counts = {}
  for (const heading of headings) {
    if (heading) {
      counts[heading] = 0
    }
  }
  return counts
}

export const sumAll = (spec: ReportSpec, filesMetricsMap: FilesMetricsMap) => {
  const counts = mkInitialCounts(spec)

  for (const filename of Object.keys(filesMetricsMap)) {
    const fileMetrics = filesMetricsMap[filename]
    for (const heading of Object.keys(fileMetrics)) {
      // metric value may be boolean, but that'll cast here to 1 or 0
      counts[heading] += fileMetrics[heading]
    }
  }

  counts[FILES_SUM_METRIC_KEY] = Object.keys(filesMetricsMap).length

  return counts
}

export const sumByOwner = (
  spec: ReportSpec,
  filesMetricsMap: FilesMetricsMap,
  ownerEntries: OwnersEntry[],
) => {
  const ownerCounts: { [owner: string]: { [heading: string]: number } } = {}

  for (const filename of Object.keys(filesMetricsMap)) {
    const owners = whoOwns(ownerEntries, filename)
    for (const owner of owners) {
      if (!ownerCounts[owner]) {
        ownerCounts[owner] = mkInitialCounts(spec)
      }
      const fileMetrics = filesMetricsMap[filename]
      for (const heading of Object.keys(fileMetrics)) {
        // HACK for passing owners info on the metrics object
        // eslint-disable-next-line no-continue
        if (heading === 'owners') continue
        // metric value may be boolean, but that'll cast here to 1 or 0
        ownerCounts[owner][heading] += fileMetrics[heading]
      }
      ownerCounts[owner][FILES_SUM_METRIC_KEY] += 1
    }
  }

  return ownerCounts
}
