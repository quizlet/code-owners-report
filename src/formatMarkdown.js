// @flow
import markdownTable from 'markdown-table'

import type { ReportSpec, FilesMetricsMap } from './config'

import { FILES_SUM_METRIC_KEY } from './config'

const metricsFor = (metricKeys: string[], record: Object) =>
  metricKeys.map(metric => record[metric] || '-')

export const markdownAllTable = (specMetricKeys: string[], all: Object) => {
  const metricKeys = [FILES_SUM_METRIC_KEY, ...specMetricKeys]

  const headers = metricKeys

  const rows = metricsFor(metricKeys, all)

  return markdownTable([headers, rows])
}

export const markdownOwnersTable = (
  specMetricKeys: string[],
  ownerSum: Object,
) => {
  const metricKeys = [FILES_SUM_METRIC_KEY, ...specMetricKeys]

  const headers = ['owner', ...metricKeys]

  const rows = Object.keys(ownerSum).map(owner => [
    owner,
    ...metricsFor(metricKeys, ownerSum[owner]),
  ])

  return markdownTable([headers, ...rows])
}

export const markdownFileTable = (
  metricKeys: string[],
  eachFile: FilesMetricsMap,
) => {
  const ownersAndMetricsKeys = ['owners', ...metricKeys]
  const headers = ['filename', ...ownersAndMetricsKeys]

  const rows = Object.keys(eachFile).map(filename => [
    filename,
    ...metricsFor(ownersAndMetricsKeys, eachFile[filename]),
  ])

  return markdownTable([headers, ...rows])
}

/**
 * Format the report as Markdown text
 */
export const formatReport = (reportSpec: ReportSpec, report: *) => {
  const specMetricKeys = [
    'owners',
    ...Object.keys(reportSpec.regexpMetrics || {}),
    ...Object.keys(reportSpec.eslintFlags || {}),
  ]

  const content = [
    'This report was autogenerated by code-owners-report.',
    '## All files',
    markdownAllTable(specMetricKeys, report.allSum),
    '## By owner (non-exclusive)',
    report.ownerSum
      ? markdownOwnersTable(specMetricKeys, report.ownerSum)
      : 'Missing CODEOWNERS',
    '## Each file',
    markdownFileTable(specMetricKeys, report.eachFile),
    '\n',
  ].join('\n\n')

  return content
}
