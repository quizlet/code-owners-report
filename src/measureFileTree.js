// @flow
import fs from 'mz/fs'
import { countRegExpMatches } from './regexpMetrics'
import { countEslintRuleViolations } from './eslintMetrics'

import type { FileMetrics, FilesMetricsMap, ReportSpec } from './config'

const measureFile = async (
  filePath: string,
  spec: ReportSpec,
): Promise<FileMetrics> => {
  const metrics = {}

  const contents = await fs.readFile(filePath, { encoding: 'utf8' })

  if (spec.regexpMetrics) {
    Object.assign(metrics, countRegExpMatches(contents, spec.regexpMetrics))
  }

  if (spec.eslintFlags) {
    Object.assign(
      metrics,
      countEslintRuleViolations(contents, spec.eslintFlags),
    )
  }

  return metrics
}

/**
 * Report metrics on every file within `dir`
 */
export default async function measureFileTree(
  dir: string,
  spec: ReportSpec,
): Promise<FilesMetricsMap> {
  console.log('Measuring directory', dir)

  const metrics: FilesMetricsMap = {}

  // consider https://github.com/jergason/recursive-readdir
  const files = fs.readdirSync(dir) || []
  // Revisit this if perf is an issue.
  // eslint-disable-next-line no-await-in-loop
  for (const file of files) {
    const subpath = `${dir}/${file}`
    if (fs.statSync(subpath).isDirectory()) {
      const treeMetrics: FilesMetricsMap = await measureFileTree(
        `${subpath}/`,
        spec,
      )
      Object.assign(metrics, treeMetrics)
    } else if (spec.omit && subpath.match(spec.omit)) {
      console.log('  omitting file', subpath)
    } else {
      console.log('  measuring file', subpath)
      const fileMetrics: FileMetrics = await measureFile(subpath, spec)
      metrics[subpath] = fileMetrics
    }
  }
  return metrics
}
