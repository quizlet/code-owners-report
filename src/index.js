// @flow

import measureFileTree from './measureFileTree'
import { parseCodeownersFile, findCodeownersPath } from './codeowners'

import { sumAll, sumByOwner } from './aggregateCounts'

import type { ReportSpec } from './config'

const sampleSpec: ReportSpec = {
  eslintRules: { 'no-console': true },
  regexpMetrics: { metrics: /metrics/ },
}

const ownerEntries = parseCodeownersFile(findCodeownersPath())

measureFileTree('src', sampleSpec).then(filesMetricsMap => {
  const all = sumAll(sampleSpec, filesMetricsMap)
  const byOwner = sumByOwner(sampleSpec, filesMetricsMap, ownerEntries)
  console.log({ byOwner, all })
})
