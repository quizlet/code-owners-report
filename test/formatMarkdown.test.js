import { markdownFileTable } from '../src/formatMarkdown'

describe('markdownFileTable', () => {
  it('renders', () => {
    const metricKeys = ['metric1', 'metric2']
    const metricsMap = {
      file1: { owners: ['alice', 'bob'], metric1: 9, metric2: true },
      file2: { owners: ['cathy', 'dan'], metric1: -1, metric2: false },
    }
    expect(markdownFileTable(metricKeys, metricsMap)).toMatchSnapshot()
  })
})
