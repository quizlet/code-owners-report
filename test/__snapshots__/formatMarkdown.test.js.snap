// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdownFileTable renders 1`] = `
"| filename | owners    | metric1 | metric2 |
| -------- | --------- | ------- | ------- |
| file1    | alice,bob | 9       | true    |
| file2    | cathy,dan | -1      | -       |"
`;
